### PR TITLE
MM-15197: config migrate SAML keys and certificates

### PR DESCRIPTION
**Pull request by therealpuneeth20**
#### Summary
This PR provides a method to migrate/copy SAML keys and certificate from one store to another store. Its used in `config migrate` command.

#### Ticket Link
  Fixes https://github.com/mattermost/mattermost-server/issues/10654

### CODE REVIEW

File: cmd/mattermost/commands/config.go [hunk 1]
```diff
@@ -254,10 +254,9 @@ func configMigrateCmdF(command *cobra.Command, args []string) error {
 		return errors.Wrap(err, "failed to read --to config")
 	}
 
-	// Copy config from source to destination
-	_, err = toConfigStore.Set(fromConfigStore.Get())
+	err = config.MigrateConfig(fromConfigStore, toConfigStore)
```
Comment (lieut-data) [inline cmd/mattermost/commands/config.go]: 👍 for pushing this into the `config` package

File: config/store.go [hunk 2]
```diff
@@ -60,3 +61,47 @@ func NewStore(dsn string, watch bool) (Store, error) {
...
+func MigrateConfig(sourceStore, destinationStore Store) error {
+	// Copy config from source to destination
+	sourceStoreConfig := sourceStore.Get()
+
+	_, err := destinationStore.Set(sourceStoreConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to migrate config")
+	}
+
+	//get the certificate files from sourceStore
```
Comment (lieut-data) [inline config/store.go]: nit: Can you make the comments uniform with others in this file?
```suggestion
	// Get the certificate files from sourceStore
```

File: config/store.go [hunk 3]
```diff
@@ -60,3 +61,47 @@ func NewStore(dsn string, watch bool) (Store, error) {
...
+	if err != nil {
+		return errors.Wrap(err, "failed to migrate config")
+	}
+
+	privateKeyFile, err := sourceStore.GetFile(*sourceStoreConfig.SamlSettings.PrivateKeyFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to migrate config")
+	}
+
+	//copy files to destinationStore
```
Comment (lieut-data) [inline config/store.go]:
```suggestion
	// Copy files to destinationStore
```

File: config/store.go [hunk 4]
```diff
@@ -60,3 +61,47 @@ func NewStore(dsn string, watch bool) (Store, error) {
...
+
+	_, err := destinationStore.Set(sourceStoreConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to migrate config")
+	}
+
+	//get the certificate files from sourceStore
+	idpCertificateFile, err := sourceStore.GetFile(*sourceStoreConfig.SamlSettings.IdpCertificateFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to migrate config")
```
Comment (lieut-data) [inline config/store.go]: Let's update the wrapped error messages here and below to call out more clearly what might have failed here -- e.g.
```suggestion
		return errors.Wrap(err, "failed to fetch SAML IDP certificate file")
```

File: config/store.go [hunk 5]
```diff
@@ -60,3 +61,47 @@ func NewStore(dsn string, watch bool) (Store, error) {
...
+	// Copy config from source to destination
+	sourceStoreConfig := sourceStore.Get()
+
+	_, err := destinationStore.Set(sourceStoreConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to migrate config")
+	}
+
+	//get the certificate files from sourceStore
+	idpCertificateFile, err := sourceStore.GetFile(*sourceStoreConfig.SamlSettings.IdpCertificateFile)
```
Comment (lieut-data) [inline config/store.go]: I believe this (and below) will fail if there is no configured IDP certificate -- let's check if these settings are empty before trying to migrate the files.

File: config/store.go [hunk 6]
```diff
@@ -60,3 +61,47 @@ func NewStore(dsn string, watch bool) (Store, error) {
 
 	return NewFileStore(dsn, watch)
 }
+
+func MigrateConfig(sourceStore, destinationStore Store) error {
```
Comment (lieut-data) [inline config/store.go]: Propose moving this into a `migrate.go` to leave `store.go` focussed on exposing the interface and constructor function.

**Comment by jasonblais** [comment]
Hey @therealpuneeth20, let us know if you have any questions about the proposed changes above? Would be glad to help!

**Comment by therealpuneeth20** [comment]
@jasonblais Sorry, I have been busy with other things. The suggested changes sounds good to me. Will push some commits over the weekend.

**Comment by jasonblais** [comment]
Sounds good, no worries! Thanks for the heads up @therealpuneeth20 👍

**Comment by hanzei** [comment]
Hey @therealpuneeth20,

Can we help with anything to allow you to complete the PR?

Please note that we will close this PR after another 10 days of inactivity. If you don't object, other contributors might complete this PR for you.

**Comment by RajatVaryani** [comment]
@hanzei Can I complete this PR?

**Comment by hanzei** [comment]
Sure @RajatVaryani. I will close this PR and you can submit a new one based on this changes. Thanks for working on this :+1: 

@therealpuneeth20 Thank you for your contribution

**Pull request closed without merging**